### ISSUE TITLE
Image Distortions should be able to be applied to batches. [Feature Request?]

### ISSUE THREAD

**Issue by jubjamie** [tensorflow/tensorflow#8926]
The tf.image distortion functions only accept a single image as an input whereas it would be much more useful to be able to place the distortion within your graph flow so any batches of images that pass through it get distorted too. I believe a current solution is as suggeested by @mrry [here on Stack Overflow](http://stackoverflow.com/questions/38920240/tensorflow-image-operations-for-batches) however it feels like, much like the other ops for images work, being able to take batches would be much more useful.

Is the solution proposed by @mrry the accepted method or should the distortion functions be taking batches?

**Comment by cancan101** [comment]
I made a similar changed to `resize_image_with_crop_or_pad `in https://github.com/tensorflow/tensorflow/pull/7369.

**Comment by jubjamie** [comment]
@cancan101 that would be great, something like that. The map_fn method is looking promising for the mean time (as i'm short on time to go and patch TF myself) but I had some PIL/Pillow tools built already in Python so might look at trying to feed in and out of that (although would much rather prefer an in-graph solution!)

**Comment by aselle** [comment]
@mrry, since you commented on the Stackoverflow, could you comment on this?

**Comment by mrry** [comment]
I think it makes sense to make the image API more consistent (in a similar manner to what @cancan101 did), but I'll defer to anyone who feels more ownership over the image API (e.g. @martinwicke, @shlens?).

**Comment by martinwicke** [comment]
See also #1029, #521. @cancan101 did #7369 fix #521?

**Comment by bhack** [comment]
And https://github.com/tensorflow/tensorflow/issues/8465?

**Comment by zo7** [comment]
Just to chime in, it seems that some image ops (e.g. `tf.image.random_brightness()`, `tf.image.random_contrast()`) allow batches, while others (e.g. `tf.image.random_saturation()`, `tf.image.random_hue()`) do not – it would make sense to either make all the image ops accept batches, or at least note on the documentation that they don't.

(also, does the "awaiting tensorflower" flag mean a TF dev is working on this, or is this open to contributions?)

**Comment by martinwicke** [comment]
Stale tags. Fixed now.

**Comment by itsmeolivia** [comment]
Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.

**Comment by bhack** [comment]
@itsmeolivia It is a little bit strange to close an issue with tag awaiting tensorflower. Tensorflower arrived to close just close the issue 😄? And now only the issue owner or a tensorflower could reopen it.

**Comment by itsmeolivia** [comment]
@bhack Hey! I don't see what's strange.  It was not awaiting a tensorflower as per the tags.  It hasn't been commented on in almost two months so my original statement, "Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you"  still makes sense, especially in light of 1.2 coming out today 😄

**Comment by bhack** [comment]
Was not commented cause we was awaiting a tensorflower tagged by a tensorflower. Cause is old I don't know if the owner (@jubjamie) is still interested on this topic. If he is not active anymore only  a tensorflow could reopen this issue.

**Comment by bhack** [comment]
@martinwicke Can we considered this to be solved by [tf.map_fn()](https://stackoverflow.com/questions/38920240/tensorflow-image-operations-for-batches)?

**Comment by martinwicke** [comment]
Nah, last comment was April 19, I removed awaiting label April 20, marked contributions welcome, no activity since then. 

map_fn is not the perfect solution, but it's a solution. If you think it's valuable to leave open as contributions welcome, we can leave it open. There's not enough demand to prioritize this though, and nobody's lining up to do it (although it would be a fairly good entry-level contribution).

**Comment by bhack** [comment]
You a right I've missed the contributions welcome label. I think that this issu was linked to many other issue about batch in image api some of that are closed in this closing run. Can we live one issue opened for all realted to batching in image api?

**Comment by bhack** [comment]
I think also an unique issue doesn't make sense cause resize_image_with_crop_or_pad in image api now support batch

**Comment by jubjamie** [comment]
I'm very much still active! However I don't have time to work on the contribution. I raised it as an inconsistency that I thought the Devs would want to fix but if it's being left as contributions only for this long with no response then I guess we'll just have to wait until someone else has the same issue but with more time than me!

**Comment by zo7** [comment]
```
>>> import tensorflow as tf
>>> tf.__version__
'1.2.0'
>>> x = tf.placeholder(tf.float32, [None, 32, 32, 3])
>>> tf.image.random_brightness(x, max_delta=0.125)
<tf.Tensor 'adjust_brightness/Identity_1:0' shape=(?, 32, 32, 3) dtype=float32>
>>> tf.image.random_contrast(x, lower=0.5, upper=1.5)
<tf.Tensor 'adjust_contrast/Identity_1:0' shape=(?, 32, 32, 3) dtype=float32>
>>> tf.image.random_flip_left_right(x)
...
ValueError: 'image' (shape (?, 32, 32, 3)) must be three-dimensional.
>>> tf.image.random_flip_up_down(x)
...
ValueError: 'image' (shape (?, 32, 32, 3)) must be three-dimensional.
>>> tf.image.random_saturation(x, lower=0.5, upper=1.5)
...
ValueError: Shape must be rank 3 but is rank 4 for 'adjust_saturation/Slice' (op: 'Slice') with input shapes: [?,32,32,3], [3], [3].
>>> tf.image.random_hue(x, max_delta=0.2)
...
ValueError: Shape must be rank 3 but is rank 4 for 'adjust_hue/Slice' (op: 'Slice') with input shapes: [?,32,32,3], [3], [3].
```
Not all image ops support batch still, can we reopen? I raised a separate issue about making the image module API more consistent (#9318) but closed it because this issue existed. This has been on the back of my mind to contribute to because it would benefit a model I have but I haven't had time to look into it recently.

**Comment by martinwicke** [comment]
Let's keep this open to track progress on this issue.

**Comment by drakerehfeld** [comment]
I know this has been briefly mentioned, but I've had great results just using ```tf.map_fn```.

```
hue = lambda x: tf.image.random_hue(x, 0.5)
hue_imgs = tf.map_fn(hue, input_imgs)
```

This seems to work on the whole batch. 

+1 to the fact that it is annoying how some ops (random_brightness) work on batches, but others (random_hue) do not :-/

**Comment by perezjln** [comment]
tf.image.random_flip_left_right() on batch would be fine indeed ...

**Comment by JoshVarty** [comment]
I would like to volunteer to fix these functions:
- `flip_left_right`
- `flip_up_down`
- `random_flip_left_right`
- `random_flip_up_down`

There are two I'm not familiar with but will also try if I think I can manage:
- `transpose_image`
- `rot90`

**Comment by yongtang** [comment]
The batch support for `adjust_hue` should have been enabled through PR #14187. And the batch support for `adjust_saturaton` will likely be addressed if PR #14794 is accepted.

**Comment by martin-gorner** [comment]
I'm on Tensorflow 1.4.1 and I still got "Shape must be rank 3 but is rank 4 for 'adjust_hue/Slice' (op: 'Slice') with input shapes: [?,256,256,3]".

**Comment by yongtang** [comment]
@martin-gorner From the commit history https://github.com/tensorflow/tensorflow/commit/d5f0634312e9426414687c0b718c931a256a082f

It seems that this commit is only part of the `v1.5.0-rc1` and `v1.5.0-rc0` (not 1.4 or 1.4.1). I think the fix will be there once 1.5 is released.

**Comment by georgesterpu** [comment]
Hi,
Would it be possible to make the distortion functions applicable not only on the entire batch, but also with a different seed per batch element ? This would lead to batches that have both the original and distorted examples, hence more diversity.

Something like:
`inputs = tf.map_fn(lambda item: tf.image.random_flip_left_right(item[1], seed=item[0]), enumerate(inputs))`
although it doesn't work this way, as there's an enum-to-tensor conversion.

I was also thinking about computing the seed as a function of the image (e.g. sum of all pixels), but it doesn't work with seed as Tensor, only int.
